### PR TITLE
 Adding Information Theoretic Rule

### LINKED ISSUE #122: Feature Addition: "Classic" and "Information Theoretic" Decision Rules
**Summary**
Pavel has had fairly clean copies of these decision rules for some time now, I'm hoping to implement them in crmPack (a) because they are a nice addition and (b) as an opportunity to familiarise myself with structure of the code. 

**References**
- O'Quigley J, Pepe M, Fisher L. Continual reassessment method: a practical design for phase 1 clinical trials in cancer. Biometrics. 1990;46(1):33-48
- Mozgunov P, Jaki T. Improving safety of the continual reassessment method via a modified allocation rule.Statistics in Medicine. 2020;39:906–922. https://doi.org/10.1002/sim.8450

**Additional Information**
I've copied over the template as a rough to do list, although I'm not sure how strictly I'll end up following it.

To do:
- [x] find correct locations for the "..." later on in this checklist
- [x] Implement the class in the file `RulesClass`
- [x] Implement function in the `Rules-Method` file
- [ ] Add tests
- [ ] Add to pkgdown.yaml

**Prototype code:**
[Check NextBestMinDist.txt](https://github.com/Roche/crmPack/files/8089314/Check.NextBestMinDist.txt)
[Check NextBestInfTh.txt](https://github.com/Roche/crmPack/files/8089316/Check.NextBestInfTh.txt)
[NextBestInfTh.txt](https://github.com/Roche/crmPack/files/8089317/NextBestInfTh.txt)
[NextBestMinDist.txt](https://github.com/Roche/crmPack/files/8089318/NextBestMinDist.txt)

### PR DESCRIPTION
**Pull request by pc494**
# Pull Request

Fixes #122 - this continues from in a fork at #149 (now closed).

### CODE REVIEW

**Review by danielinteractive — COMMENTED** [review]
thanks Phillip, looks good so far, can you add the tests?

**Comment by pc494** [comment]
> thanks Phillip, looks good so far, can you add the tests?

Yes, that's next. Sorry for the bite sized chunks on this, I had been hoping that these changes would mean the the CI would run the tests and coverage for me so that I could see what was going on...

File: R/Rules-methods.R [hunk 1]
```diff
@@ -664,8 +664,7 @@ setMethod("nextBest",
 ## ============================================================
 
 h_info_theory_dist <- function(probability,target,asymmetry){
-  delta = ((probability-target)^2)/((probability^asymmetry)*(1-probability)^(2-asymmetry)))
-  return delta
+  return ((probability-target)^2)/(((probability^asymmetry)*(1-probability)^(2-asymmetry)))
```
Comment (danielinteractive) [inline R/Rules-methods.R:667]:
```suggestion
  ((probability-target)^2)/(((probability^asymmetry)*(1-probability)^(2-asymmetry)))
```
(no explicit return needed in R , and style wise it is better not to do it)

**Comment by danielinteractive** [comment]
@dinakar29 do you have any idea why the Coverage reports errors here? since unit tests are passing. it is not that important for us at this point, more curious.

File: R/Rules-methods.R (hunk 1 above)
Comment (pc494) [reply R/Rules-methods.R:667]: old habits, will fix.

**Comment by dinakar29** [comment]
I was able to reproduce this locally, and upon reviewing the `testthat.Rout.fail` file, I see this as being the root cause:

```R
Backtrace:
    ▆
 1. ├─crmPack::mcmc(data = data, model = model, options = options) at test-Model-class.R:181:2
 2. └─crmPack::mcmc(data = data, model = model, options = options)
 3.   └─crmPack .local(data, model, options, ...)
 4.     └─rjags::jags.model(...)
── Error (test-Model-class.R:207:3): MCMC computes correct values for ProbitLogNormalRel model ──
Error in `rjags::jags.model(file = model_file, data = model_data, inits = c(model_inits,
    .RNG.name = h_null_if_na(options@rng_kind), .RNG.seed = h_null_if_na(options@rng_seed)),
    quiet = !is_logging_enabled(), n.adapt = 0)`:
Error parsing model file:
syntax error on line 3 near "{"
```

Not sure where or how to resolve this immediately, but as an enhancement to the coverage workflow, I'll update it so we can view these types of errors in the CI logs.

**Comment by danielinteractive** [comment]
@pc494 let me know if you would like to chat about tests etc

**Comment by pc494** [comment]
Thanks @danielinteractive, I think I've got the syntax down and am going to bug Pavel for a sensible test case.

File: tests/testthat/test-Rules-methods.R [hunk 2]
```diff
@@ -329,3 +329,18 @@ test_that("StoppingLowestDoseHSRBeta works correctly if first active dose is not
   )
   expect_identical(result, expected) # First active dose not applied.
 })
+
+test_that("NextBestInfTheory can be initiliazed ", {
+  # set up
+  newMyNextBest <- NextBestInfTheory(target = 0.25, asymmetry=0.1)
```
Comment (danielinteractive) [inline tests/testthat/test-Rules-methods.R:335]:
```suggestion
  new_my_next_best <- NextBestInfTheory(target = 0.25, asymmetry = 0.1)
```

File: tests/testthat/test-Rules-methods.R [hunk 3]
```diff
@@ -329,3 +329,18 @@ test_that("StoppingLowestDoseHSRBeta works correctly if first active dose is not
...
+})
+
+test_that("Helper function (h_info_theory_dist) for NextBestInfTheory is correct", {
+  # values calculated using a different program
+  p <- c(0.01, 0.2, 0.7)
+  gamma <- c(0.5, 0, 0.3)
+  a <- c(1, 1.8, 0.5)
+  expected <- c(24.25, 0.76, 1.16)
+  calculated <- h_info_theory_dist(p, gamma, a)
+  expect_equal(calculated, expected, tolerance=0.01)
```
Comment (danielinteractive) [inline tests/testthat/test-Rules-methods.R:345]:
```suggestion
  expect_equal(calculated, expected, tolerance = 0.01)
```

**Review by danielinteractive — COMMENTED** [review]
Hi @pc494 please see my proposals to fix linter problems.

**Review by danielinteractive — COMMENTED** [review]
maybe one additional test that the `nextBest()` method result using the new class is as expected would be good.

File: tests/testthat/test-Rules-methods.R (hunk 2 above)
Comment (pc494) [reply tests/testthat/test-Rules-methods.R:335]: oh! that's what it didn't like, I got so hung up on the brackets I completely missed this, will fix.

**Comment by danielinteractive** [comment]
@pc494 let me know if you need help with test that the nextBest() method result is correct

**Comment by pc494** [comment]
Okay, the testing now fails locally as it's calling `rbinom` with invalid arguments somewhere. I'm looking into it.

**Comment by wwojciech** [comment]
Hey @pc494 , I would be happy to help you with that if you think it makes sense. Tx!

**Comment by wwojciech** [comment]
@pc494 , @danielinteractive - FYI
I fixed the error with the last unit tests that Phillip mentioned about, together with some other linter and cosmetic issues.

File: R/Rules-methods.R [hunk 4]
```diff
@@ -663,6 +663,39 @@ setMethod("nextBest",
 
 ## ============================================================
```
Comment (danielinteractive) [inline R/Rules-methods.R:665]: sorry I forgot to mention this earlier, but can you please add documentation to this function, too? thx!

File: tests/testthat/test-Rules-methods.R [hunk 5]
```diff
@@ -329,3 +329,65 @@ test_that("StoppingLowestDoseHSRBeta works correctly if first active dose is not
   )
   expect_identical(result, expected) # First active dose not applied.
 })
+
+test_that("NextBestInfTheory can be initiliazed ", {
+  expect_silent(NextBestInfTheory(target = 0.25, asymmetry = 0.1))
+})
+
+test_that("Helper function (h_info_theory_dist) for NextBestInfTheory is correct", {
```
Comment (danielinteractive) [inline tests/testthat/test-Rules-methods.R:337]:
```suggestion
test_that("h_info_theory_dist works as expected", {
```

File: tests/testthat/test-Rules-methods.R [hunk 6]
```diff
@@ -329,3 +329,65 @@ test_that("StoppingLowestDoseHSRBeta works correctly if first active dose is not
...
+  # Values calculated using a different program.
+  p <- c(0.01, 0.2, 0.7)
+  gamma <- c(0.5, 0, 0.3)
+  a <- c(1, 1.8, 0.5)
+  expected <- c(24.25, 0.76, 1.16)
+  calculated <- h_info_theory_dist(p, gamma, a)
+  expect_equal(calculated, expected, tolerance = 0.01)
+})
+
+test_that("NextBestInfTheory produces consistent results", {
```
Comment (danielinteractive) [inline tests/testthat/test-Rules-methods.R:347]: this is nice in the sense that it looks at behavior inside study simulations - from a unit testing perspective we still need a test that looks at a single data set and what the `nextBest` method gives as result though (see my previous comment)

**Review by danielinteractive — COMMENTED** [review]
thx @pc494 (and @wwojciech for helping) - see inside last comments, in addition can you please add new class and helper function to the `pkgdown.yaml` file? thx!

**Comment by pc494** [comment]
> thx @pc494 (and @wwojciech for helping) - see inside last comments, in addition can you please add new class and helper function to the `pkgdown.yaml` file? thx!

I was able to add the helper, but I'm not sure where the "new class" would go in the `.yaml`

File: tests/testthat/test-Rules-methods.R (hunk 6 above)
Comment (danielinteractive) [reply tests/testthat/test-Rules-methods.R:347]: @pc494 can you please add another test as mentioned here?

**Comment by wwojciech** [comment]
@pc494 - The errors appearing at `R CMD Check` stage are due to missing entries in the manual of your helper function (no @param) - as we discussed.

**Comment by danielinteractive** [comment]
@pc494 @wwojciech can you maybe meet again or so to get this PR merged? I am a bit worried that this PR is 2 months in the works already. thanks!

**Comment by wwojciech** [comment]
> @pc494 @wwojciech can you maybe meet again or so to get this PR merged? I am a bit worried that this PR is 2 months in the works already. thanks!

Sure @danielinteractive, @pc494 - please let me know (on slack or via email), if you would like to discuss any blocking points.

**Comment by danielinteractive** [comment]
thx @pc494 please let me once you are ready for re-review of this PR, thank you

**Comment by pc494** [comment]
> thx @pc494 please let me once you are ready for re-review of this PR, thank you

@wwojciech and I have a meeting tomorrow afternoon after which I (fingers crossed) think this will be finished and ready to review. Will tag you to confirm.

**Review by danielinteractive — COMMENTED** [review]
Thanks @pc494 , looks good regarding automatic checks are passing now!
But one thing still open: see my previous comment about adding a test for the `nextBest()` method itself. Currently you just have tests for the whole design simulations, but on the "unit" of the next best dose calculation. Can you please still add that?

**Comment by wwojciech** [comment]
> Thanks @pc494 , looks good regarding automatic checks are passing now! But one thing still open: see my previous comment about adding a test for the `nextBest()` method itself. Currently you just have tests for the whole design simulations, but on the "unit" of the next best dose calculation. Can you please still add that?

Hi @pc494 
What Daniel means in his last response is that you would need to test `nextBest()` method for object of class `NextBestInfTheory`. Only a few simple tests for this would be needed. See e.g. how `dose()` (or `prob()`) method is tested for object of many different model classes in `test-Model-methods.R` file.

**Comment by pc494** [comment]
Thanks for the help both, hoping I should be able to deal with this.

**Comment by pc494** [comment]
@wwojciech 

I think I've got most of this down now:

```r
dlt_model <- NextBestInfTheory(target=0.25,asymmetry=0.1)
# confused
samples <- h_as_samples(list(target = c(0.4, 0.5), asymmetry = c(1, 0.6)))
# think this should work once I have samples down
prob_result <- prob(c(50, 75), dlt_model,samples)
dose_result <- dose(c(0.2, 0.4), dlt_model,samples)

expect_snapshot(prob_result)
expect_snapshot(dose_result)
```

apart from the sample argument, which I still don't really understand. Is there a helper I can use for this?

**Comment by wwojciech** [comment]
Hi @pc494, so actually this is not entirely correct. It should be something like:
```                     
nb_it <- NextBestInfTheory(target=0.25, asymmetry=0.1)
samples <- samples <- h_as_samples(<some_list_with_samples>)
model <- <some_model>
data <- <some_data>

result <- nextBext(nextBest = nb_it, doselimit = <some_limit>, samples = samples, model = model, data = data)
expected <- <expected_output_from_above_nextBest>

expect_identical(result, expected)
```

**Comment by pc494** [comment]
Thanks @wwojciech, I hope this will pass now.

**Comment by pc494** [comment]
The test that failed on the CI passed locally so I'm not sure how to go about debugging.

**Comment by pc494** [comment]
I think this ready for review again.

**Review by danielinteractive — APPROVED** [review]
Thanks @pc494 and @wwojciech !

**Pull request merged**